### PR TITLE
netcdf-c: major refactoring, new variants and versions

### PR DESCRIPTION
**Pull request by skosukhin**
This is a major update for the `netcdf-c` package. I don't think that we can or should support new versions of the library before we merge this.

Brief summary of changes (see commit messages for the full one):
1. General refactoring of the package.
2. Consistent handling of the compiler and linker flags (improves static linking scenarios).
3. Added missing dependencies and removed (constrained) the redundant ones.
4. New variants `szip` and `blosc` to enable/disable the respective compression plugins and consistent handling of the rest of the plugins (all compression plugins are enabled by default).
5. New variant `nczarr_zip` to enable/disable the zipfile format storage (disabled by default; NCZarr format itself is unconditionally enabled starting version `4.8.1`, which is similar to how we currently treat HDF5).
6. New variant `byterange` to enable/disable the byte-range I/O (disabled by default).
7. The patches are fetched from the upstream repo whenever possible.
8. New version `4.9.2`.

I understand that this PR modifies a lot of things and it might be difficult to review it but the commit history is rather clean and almost every commit in it can become a separate PR if it helps.

Also, I am not ready to support the `CMake` builder yet (I have only removed a single obviously redundant line of code there). Therefore, the new variants might not work as expected when `build_system=cmake` and especially when `platform=windows`.

### CODE REVIEW

**Comment by alalazo** [comment]
@skosukhin Apologies, but clicking "Rebase the PR" to trigger tests run resulted in this mess somehow :worried: So, can you please push the branch again and ping me for a review?

EDIT: otherwise I can also force push on your branch, if you'd like me to.

**Comment by alalazo** [comment]
@skosukhin Fixed it manually, I rebased and push forced to your branch - since permission was granted for it.

**Review by alalazo — DISMISSED** [review]
LGTM as far as Spack idioms are concerned. Waiting for @WardF review. Ping me in a week or so if there's no reply and I'll merge, provided that pipelines pass too.

File: var/spack/repos/builtin/packages/netcdf-c/package.py
Comment (johnwparent) [inline var/spack/repos/builtin/packages/netcdf-c/package.py]: This should be `with when("platform!=windows")` as we should be referring to the spec target platform. But since we don't have a `!=` for specs, the current convention is to express this as:
```
    for plat in ["darwin", "cray", "linux"]:
        with when("platform=%s" % plat):
```

**Comment by skosukhin** [comment]
The compiler wrapper filtering doesn't work (see #36689). In addition to that, [`_filter_compiler_wrappers_impl`](https://github.com/spack/spack/blob/385834bd43e1958667b7f75074e07f412d6f58ab/lib/spack/spack/mixins.py#L62) cannot be used for the builder classes but only for the packages.

I have considered three options to circumvent the issue:
1. Patch `_filter_compiler_wrappers_impl`:
    ```patch
    --- a/lib/spack/spack/mixins.py
    +++ b/lib/spack/spack/mixins.py
    @@ -62 +62,3 @@ def filter_compiler_wrappers(*files, **kwargs):
    -    def _filter_compiler_wrappers_impl(self):
    +    def _filter_compiler_wrappers_impl(package_or_builder_self):
    +        self = package_or_builder_self.spec.package
    ```
    and call `filter_compiler_wrappers` inside the builder classes.
2. Copy and adjust the relevant part of `filter_compiler_wrappers` to `netcdf-c/package.py` and call it as additional `@run_after("install")` callbacks in both builder classes.
3. Temporarily remove the builder classes, put both `cmake_args` and `configure_args` into the package class and rely on the automatically generated wrapper class.

I decided to go for the last option for now. Mainly because I don't want to touch the core files and/or introduce code duplication. It seems to work but I might be missing some details.

@alalazo please, take a look. I also think that we have waited long enough to feel free to merge this as soon as the tests pass. Unfortunately, there are problems with `audit` and rebasing didn't help.

File: var/spack/repos/builtin/packages/netcdf-c/package.py
Comment (skosukhin) [reply var/spack/repos/builtin/packages/netcdf-c/package.py]: @johnwparent ok, you have already corrected your message. I don't mind this suggestion. I just took this condition from [your code](https://github.com/spack/spack/blob/385834bd43e1958667b7f75074e07f412d6f58ab/var/spack/repos/builtin/packages/netcdf-c/package.py#L103) and could not find anything similar to `spack.compilers.supported_compilers` (which we use in similar cases), but for the platforms. I now see that this pattern is used in many packages, thank you.

Since you are here, maybe you can comment regarding the [`4.8.1-win-hdf5-with-zlib.patch`](https://github.com/spack/spack/pull/36485/files#diff-9bb18f61f441605ae314f7ccaa7feebb323fe12be6df13a0364f93903977477cR57-R60)?

Comment (johnwparent) [reply var/spack/repos/builtin/packages/netcdf-c/package.py]: Yes sorry, I autopiloted to hit `comment` before I was done actually writing my comment. My code was also not ideal/correct and I should have caught that in review.
IIRC there's something in the works for the platform equivalent of `spack.compilers.supported_compilers` but is not yet ready for deployment/is still preliminary. The pattern is unwieldy and verbose but is the convention at the moment. 

In general, if you're modifying something that impacts Windows support, absolutely feel free to summon me if you feel the need/are unsure as there is very minimal CI protecting the Windows side of things.

The patch was necessary for me to get this package building against hdf5 with version 4.8.1. IIRC there was an issue raised either on the netcdf-c or hdf5 repos that covered this exact problem and is still open. If you'd like I can try to track that down.

**Comment by alalazo** [comment]
@skosukhin I think I just fixed the issue with `audit`, see https://community.codecov.com/t/codecov-yanked-from-pypi-all-versions/4259 for more information. If you rebase CI should work.

File: var/spack/repos/builtin/packages/netcdf-c/package.py
Comment (skosukhin) [reply var/spack/repos/builtin/packages/netcdf-c/package.py]: Yes, I would like to know more about the patch. As far as I understand, `H5public.h` includes `H5pubconf.h` and the expected `H5_HAVE_ZLIB_H` macro is supposed to be set anyway. I am also curious whether the patch is needed for later versions of NetCDF, which also include `H5public.h`.

**Comment by alalazo** [comment]
> I decided to go for the last option for now. Mainly because I don't want to touch the core files and/or introduce code duplication. It seems to work but I might be missing some details

Please don't do that, it's a use of the classes that we don't support or test (if there are multiple build systems, packages should use builders). I think I fixed the issue with compiler wrappers in #36825 If so we can proceed from there.

File: var/spack/repos/builtin/packages/netcdf-c/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/netcdf-c/package.py:116]: (not a request, just a comment) Here there is potential for a multi-valued variant, if it seems a good idea.

File: var/spack/repos/builtin/packages/netcdf-c/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/netcdf-c/package.py:143]: Why the double underscore? I need to review if that obscure name mangling Python feature kicks in here...

**Review by alalazo — DISMISSED** [review]
Skimming through the recipe, it looks good to me on Spack idioms. I trust the maintainer more than myself for the semantic of the changes. Only a couple minor questions.

File: var/spack/repos/builtin/packages/netcdf-c/package.py
Comment (skosukhin) [reply var/spack/repos/builtin/packages/netcdf-c/package.py:116]: I didn't go for a multi-valued variant because:
1. The `zstd` variant already exists and I didn't want this PR to be doomed like #32476 where I tried to touch the existing variants and it has never been accepted.
2. I couldn't come up with a good name and a set of values for a multi-valued variant. If the name is `plugins` and the possible values are `none`, `szip`, `blosc` and `zstd`, then its unclear how to interpret `plugins=none`. On one hand, that can mean "do not install plugins at all", which is possible. On the other hand, it can be "do not install szip, blosc, zstd but install the two that are supposed to be ["always available"](https://github.com/Unidata/netcdf-c/blob/b30b4e87cff202ab3c6511ade05a58019fc79796/configure.ac#L1947-L1948). However, I can put more thinking into it if you say that the first reason is irrelevant.

Comment (alalazo) [reply var/spack/repos/builtin/packages/netcdf-c/package.py:116]: Not a worry for this PR. If you do this modification in another PR[^1], ping me (repeatedly if I seem to miss the PR :slightly_smiling_face: )

[^1]: that is to say, if you think it's a modification worth doing!

File: var/spack/repos/builtin/packages/netcdf-c/package.py
Comment (skosukhin) [reply var/spack/repos/builtin/packages/netcdf-c/package.py:143]: The first time I tried to create a loop in the package scope didn't go well (see [here](https://github.com/spack/spack/blob/32808f4fb51196634a6529e75ce2d81a7e3de229/var/spack/repos/builtin/packages/knem/package.py#L53-L57)). It took me a while to understand that I should not use `for compiler in` there. Since then, I try to name variables in this scope so that they have minimal chance to collide with something else. Later, I realised that the variable remain with the class instances and started explicitly removing them.

Comment (johnwparent) [reply var/spack/repos/builtin/packages/netcdf-c/package.py:143]: @alalazo the name mangling this would effect this variable, but only outside of this scope, so while this does eschew PEP conventions it should be fine considering we don't need to access `__p` outside of this scope. That being said, @skosukhin is there some reason something like `plat` or `spack_host_platform` or something more within python naming conventions wouldn't work?

For reference/convenience: https://peps.python.org/pep-0008/#method-names-and-instance-variables

**Comment by johnwparent** [comment]
@alalazo I see you approved but can we hold off on merging here? This PR breaks the package in a few different ways on Windows and it'll take me a moment to diagnose and come up with a fix.

**Comment by johnwparent** [comment]
@skosukhin How would you like to coordinate to on fixes to the Windows side of things here? I've got a patch that fixes one of the issues I've uncovered, but I'm still digging on the rest. I can push directly to your branch if I have permissions/you're ok with that, or I can upload patches here or make suggestions in the UI you can apply?

File: var/spack/repos/builtin/packages/netcdf-c/package.py
Comment (alalazo) [reply var/spack/repos/builtin/packages/netcdf-c/package.py:143]: @skosukhin Sure, I agree with that. Deleting is zealous, but not wrong. I was just wondering why the double underscore instead of a single one :slightly_smiling_face:

**Comment by skosukhin** [comment]
@johnwparent Are the problems related to the new version `4.9.2`? Does removing it "fix" the installation on Windows? If so, I can remove the new version from this PR and you can add it in a follow-up PR together with all the required fixes. I would prefer a follow-up PR from you in any case, to be honest, but uploading patches here or making suggestions in the UI will do as well.

**Comment by johnwparent** [comment]
> @johnwparent Are the problems related to the new version `4.9.2`? Does removing it "fixes" the installation on Windows? If so, I can remove the new version from this PR and you can add it in a follow-up PR together with all the required fixes. I would prefer a follow-up PR from you in any case, to be honest, but uploading patches here or making suggestions in the UI will do as well.

@skosukhin No, not all the problems are related to the new version, although I can't say which are because I can't get through a single build of any version right now. The current issues seem related to the new variants/dependencies you've added. The first is that the CMake builder was not updated with the appropriate cmake args to handle the `byterange` variant (which is the patch I have locally). The issue I'm currently running into is that there is a mix of MSVC runtime libraries introduced by the linkage between netcdfc and bzip2, which is preventing me from building any version and testing between `4.9.2` and `4.8.1`. This may be a bzip2 problem as well, but it seems as though it's netcdf that's bringing in the static runtime during a shared link, although perhaps the blame for that falls on bzip2 for linking to the shared runtime as a static library.

File: var/spack/repos/builtin/packages/netcdf-c/package.py
Comment (johnwparent) [inline var/spack/repos/builtin/packages/netcdf-c/package.py:307]:
```suggestion
            self.define_from_variant("BUILD_SHARED_LIBS", "shared"),
            self.define_from_variant("ENABLE_BYTERANGE", "byterange"),
```

**Comment by skosukhin** [comment]
A quick solution for bzip2 could be to make `bzip2` a dependency on non-Windows platforms only and use the vendored one on Windows for the time being.

**Comment by johnwparent** [comment]
> A quick solution for bzip2 could be to make `bzip2` a dependency on non-Windows platforms only and use the vendored one on Windows for the time being.

I'm more than happy to remove the dep on bzip2 for Windows for the time being until I have more time to sort that out, but that means disabling the use of bzip2 in netcdf-c entirely for Windows as there is no guarantee that the package will be available to the build. Windows does not vendor compression utilities, only tar.
In general, from what I understand, Spack should not be relying on a dependency "just being available on the system" arbitrarily without some sort of modeling/representation in Spack, and from a maintenance standpoint, that feels incredibly fragile/will lead to headaches/issues.

**Comment by skosukhin** [comment]
> Windows does not vendor compression utilities

But Netcdf-c [does](https://github.com/Unidata/netcdf-c/blob/main/plugins/bzlib.h).

**Comment by johnwparent** [comment]
> > Windows does not vendor compression utilities
> 
> But Netcdf-c [does](https://github.com/Unidata/netcdf-c/blob/main/plugins/bzlib.h).

Nice, very useful! I wish more libraries did this, would make my life on Windows far easier. In that case, yeah, lets disable an external bzip2 for Windows, but can we get some modeling in Spack indicating that we're using this? For example VTK/Paraview/CMake have a variant for the use of the dependencies they also vendor, as perhaps it would also be useful on unix to have this.

Something like:

```
variant("internal_bzip2", default=sys.platform == "win32")
variant("internal_<some other dep>", ...)
```

**Comment by skosukhin** [comment]
The usage of vendored libraries that are already spackaged is discouraged as well as the variants that enable it. And I agree with that.

**UPD:** However, I'm fine with using the vendored bzip2 on Windows as a temporary workaround.

**Comment by skosukhin** [comment]
And I'm sorry but in my opinion, by making the decision to build libraries from sources on Windows you chose against having easy life :)

**Comment by johnwparent** [comment]
> The usage of vendored libraries that are already spackaged is discouraged as well as the variants that enable it. And I agree with that.

There is already precedent for exceptions to this idiom for reasons that parallel this one. Don't define the other variants, but if we're making this choice IMO it should be clear to the users. Make the variant only available on Windows if need be, but if builds fails because of a bz2 it should be immediately clear where it's coming from as the default assumption would generally be a spack derived bz2. Maybe I'm completely off base here.

FWIW I'm likely just going to contribute a CMake system for bz2 as the issues are fairly fundamental (there's no shared option on Windows right now whatsoever) so this fix is going to have to hold up for a while.

Edit:
> And I'm sorry but in my opinion, by making the decision to build libraries from sources on Windows you chose against having easy life :)

_Someone_ has to build from source to produce the binaries for everyone else, no reason why that process shouldn't be reproduceable...

**Comment by alalazo** [comment]
> Make the variant only available on Windows if need be

If we need a variant, let's make it Windows only. In general, I'd avoid to have platform specific variants creep into other platforms.

**Comment by johnwparent** [comment]
> > Make the variant only available on Windows if need be
> 
> If we need a variant, let's make it Windows only. In general, I'd avoid to have platform specific variants creep into other platforms.

Makes sense/I agree.

**Comment by skosukhin** [comment]
@johnwparent please, check whether it works now.

As for the platform-specific variant that enables the usage of a vendored library, I don't agree with that but I'm not the one with the merge button here :)

**Comment by johnwparent** [comment]
> @johnwparent please, check whether it works now.
> 
> As for the platform-specific variant that enables the usage of a vendored library, I don't agree with that but I'm not the one with the merge button here :)

Running the build now.

I also do not have any merge power so I guess someone else gets to make that decision for us! Sorry @alalazo.

**Comment by johnwparent** [comment]
@skosukhin The build itself succeeds for both the new version added by this PR and the old version! However the compiler wrapper filters fail on Windows with the stack trace: 
```
  File "C:\spack_win\spack\lib\spack\spack\mixins.py", line 104, in _filter_compiler_wrappers_impl
    x.filter(wrapper_path, compiler_path, **filter_kwargs)
  File "C:\spack_win\spack\lib\spack\llnl\util\filesystem.py", line 421, in filter
    return filter_file(
  File "C:\spack_win\spack\lib\spack\llnl\util\filesystem.py", line 324, in filter_file
    unescaped = repl.replace(r"\\", "\\")
AttributeError: 'NoneType' object has no attribute 'replace'
```
FWIW I disabled the compiler wrapper filter on the HDF5 package, so we may want to do something like that here. We don't use the Spack compiler wrappers on Windows at the moment anyway, so it's a no-op either way.

File: var/spack/repos/builtin/packages/netcdf-c/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/netcdf-c/package.py:135]: This will bite us at some point, when the argument is stored as is by some object :worried: So far they are string and they get transformed to specs internally, so it's fine. I wouldn't try to encourage people to use this idiom.

**Review by alalazo — APPROVED** [review]

**Comment by alalazo** [comment]
@johnwparent Fine with you to merge this, and have a follow up later?

**Comment by johnwparent** [comment]
> @johnwparent Fine with you to merge this, and have a follow up later?

I'm on mobile at the moment so I can't tell if my compiler wrappers comments have been addressed. If they have, yes, LGTM!
Otherwise, considering that it breaks the package for windows users no, we should not merge this.

**Comment by alalazo** [comment]
@johnwparent I don't see changes to `filter_compiler_wrappers` in this PR. The last change was in https://github.com/spack/spack/pull/36825 which effectively fixed a bug that was introduced in https://github.com/spack/spack/pull/34935 and broke the build on Linux. 

So to summarize:
1. If current develop works on Windows and this breaks on filter_compiler_wrappers I'd like to investigate why.
2. Otherwise I'd just merge this, since it doesn't touch this aspect.

In case we can make `filter_compiler_wrapper` a no-op on Windows in a small follow up PR. Let me know if you have more details on the failure on Windows.

**Comment by johnwparent** [comment]
Looks like netcdf-c is already broken in develop (yet another argument for getting the Windows CI on it's feet) so this PR looks great, apologies for the false alarm, as you can see it's bit of a can of worms to keep things functional and stable on Windows at the moment.

LGTM!

**Review by johnwparent — APPROVED** [review]
Looks great now! Thanks for bearing with me/dealing with Windows!

LGTM

**Comment by vanderwb** [comment]
@skosukhin - apologies if this isn't the correct place to ask this... Is there a reason you no longer link to libcurl when `+dap` (but instead only provide library search dirs)? This change seems to be breaking my builds of netcdf-c when using +mpi. I get the following config error, which suggests contamination from the system libs:
```
configure:4614: checking whether we are cross compiling
configure:4622: /opt/cray/pe/mpich/8.1.25/ofi/gnu/9.1/bin/mpicc -o conftest   -L/glade/derecho/scratch/vanderwb/spack-tests/derecho/23.05/envs/build/opt/__spack_path_placeh
older__/__spack_path_placeholder__/_/hdf5/1.12.2/cray-mpich/8.1.25/gcc/12.2.0/lib -L/glade/derecho/scratch/vanderwb/spack-tests/derecho/23.05/envs/build/opt/__spack_path_pl
aceholder__/__spack_path_placeholder__/_/zlib/1.2.13/gcc/7.5.0/lib -L/glade/derecho/scratch/vanderwb/spack-tests/derecho/23.05/envs/build/opt/__spack_path_placeholder__/__s
pack_path_placeholder__/_/libszip/2.1.1/gcc/7.5.0/lib -L/glade/derecho/scratch/vanderwb/spack-tests/derecho/23.05/envs/build/opt/__spack_path_placeholder__/__spack_path_pla
ceholder__/_/bzip2/1.0.8/gcc/12.2.0/lib -L/glade/derecho/scratch/vanderwb/spack-tests/derecho/23.05/envs/build/opt/__spack_path_placeholder__/__spack_path_placeholder__/_/z
std/1.5.5/gcc/7.5.0/lib -L/glade/derecho/scratch/vanderwb/spack-tests/derecho/23.05/envs/build/opt/__spack_path_placeholder__/__spack_path_placeholder__/_/c-blosc/1.21.1/gc
c/12.2.0/lib -L/glade/derecho/scratch/vanderwb/spack-tests/derecho/23.05/envs/build/opt/__spack_path_placeholder__/__spack_path_placeholder__/_/curl/8.0.1/gcc/7.5.0/lib con
ftest.c  >&5
configure:4626: $? = 0
configure:4633: ./conftest
./conftest: symbol lookup error: /usr/lib64/libssh.so.4: undefined symbol: EVP_KDF_CTX_new_id, version OPENSSL_1_1_1d
```

If I change the package.py as follows, the build works without issue:
```
@@ -467,7 +467,8 @@ def configure_args(self):
         if self.spec.satisfies("@:4.7~dap+byterange"):
             extra_libs.append(self.spec["curl"].libs)
         elif "+dap" in self.spec or "+byterange" in self.spec:
-            lib_search_dirs.extend(self.spec["curl"].libs.directories)
+            extra_libs.append(self.spec["curl"].libs)
```

**Pull request merged**
